### ISSUE TITLE
Socket::recvfrom() throws an exception in non-blocking mode when there is no data to be read

### ISSUE THREAD

**Issue by dktapps** [krakjoe/pthreads#809]
Environment
==========

  * PHP: 3.1.7dev
  * pthreads: https://github.com/krakjoe/pthreads/commit/f3c1be543465e651f99284f5709e3a6b55ccf881
  * OS: windows x64

Summary
======
`socket_recvfrom()` simply emits a FALSE return when there is no data to be read (without any warning). Throwing an exception here in pthreads seems unnecessarily excessive.

In an application where nonblocking sockets are being recvfrom()d many times per second, spewing exceptions whenever there isn't data available is undesirable. Additionally, it makes it problematic to tell the difference between no-data and an actual error.

This difference in behaviour is also causing problems with tests (due to differing error messages for EAGAIN).

This is similar to the issue described in #733 .

Reference: https://github.com/php/php-src/blob/master/ext/sockets/tests/socket_sentto_recvfrom_ipv4_udp.phpt#L24

Reproducing Code
==============
```php
$socket = new \Socket(\Socket::AF_INET, \Socket::SOCK_DGRAM, \Socket::SOL_UDP);
$socket->setBlocking(false);
$socket->bind('127.0.0.1', 19132);

var_dump($socket->recvfrom($buf, 65535, 0, $addr, $port));
```

Expected Output
=============
```
bool(false)
```

Actual Output
============
```
Fatal error: Uncaught RuntimeException: Error (10035): A non-blocking socket operation could not be completed immediately.
 in pthreads-sockets-stuff\run.php:28
Stack trace:
#0 pthreads-sockets-stuff\run.php(28): Socket->recvfrom(NULL, 65535, 0, NULL, NULL)
#1 {main}
  thrown in pthreads-sockets-stuff\run.php on line 28
```

**Comment by dktapps** [comment]
Looks like php-src has a check for this in the `PHP_SOCKET_ERROR` macro: https://github.com/php/php-src/blob/master/ext/sockets/php_sockets.h#L87

**Comment by SirSnyder** [comment]
Nice to see this topic. Thought about this earlier today while I was adjusting the tests. I share your view. In some cases the api should just return false and not throw exceptions.  I will prepare a branch.

**Comment by dktapps** [comment]
This is resolved by the merge of #819 .
